File: src/deviationFunctions.cpp
#include <Rcpp.h>
using namespace Rcpp;


double KSTestC(LogicalVector indexSelection, NumericVector indexMapReferenceAttribute){
  double cumDistSubspace = 0;
  double maxDistance = 0;
  int numRemaining = sum(indexSelection);
  int numTotal = indexSelection.size();

  for(int i=0; i < numTotal; i++){
    if(indexSelection[indexMapReferenceAttribute[i]-1]){
      cumDistSubspace += 1.0/numRemaining;
    }
    double currentDiff = std::abs(((i + 1.0)/numTotal) - cumDistSubspace);
    if(currentDiff > maxDistance){
      maxDistance = currentDiff;
    }
  }
  return maxDistance;
}

LogicalVector randomSubspaceSliceC(NumericMatrix indexMap, NumericVector subspace, double alpha, int referenceDim) {

  int sliceSize = std::ceil(indexMap.rows() * (std::pow(alpha,(1.0/(subspace.size()-1)))));
  LogicalVector selectionAll(indexMap.rows());
  selectionAll.fill(1);

  NumericVector::iterator it;
  for(it = subspace.begin(); it != subspace.end(); ++it){
    if(*it != referenceDim){
      int l = std::rand() % (indexMap.rows() - sliceSize);
      int r = l + sliceSize;
      for(int j=0; j< l; j++){
        selectionAll[indexMap(j,*it - 1) -1] = 0;
      }
      for(int j=r; j< indexMap.rows(); j++){
        selectionAll[indexMap(j,*it - 1) -1] = 0;
      }
    }
  }
  return selectionAll;
}

double averageDeviationC(NumericMatrix indexMap, NumericVector subspace, double alpha, int referenceDim, int numRuns){
  double result = 0;
  for(int i=0; i<numRuns; i++){
    result += KSTestC(randomSubspaceSliceC(indexMap, subspace, alpha, referenceDim), indexMap.column(referenceDim - 1));
  }
  return result/numRuns;
}

NumericVector deviationVectorC(NumericMatrix indexMap, NumericVector subspace, double alpha, int referenceDim, int numRuns){
  NumericVector result(numRuns);
  for(int i=0; i<numRuns; i++){
    result[i] = KSTestC(randomSubspaceSliceC(indexMap, subspace, alpha, referenceDim), indexMap.column(referenceDim - 1));
  }
  return result;
}

List deviationStatisticsC(NumericMatrix indexMap, double alpha, int numRuns){
  NumericMatrix outAverage(indexMap.ncol(), indexMap.ncol());
  colnames(outAverage) = colnames(indexMap);
  rownames(outAverage) = colnames(indexMap);

  NumericMatrix outSD(indexMap.ncol(), indexMap.ncol());
  colnames(outSD) = colnames(indexMap);
  rownames(outSD) = colnames(indexMap);

  for(int i=0; i< outAverage.ncol(); i++){
    for(int j=0; j< outAverage.ncol(); j++){
      if(i==j){
        outAverage(i,j) = 0;
      }
      else{
        NumericVector subspace = NumericVector::create(i+1,j+1);
        NumericVector deviatioVector = deviationVectorC(indexMap, subspace, alpha, i+1, numRuns);
        outAverage(i,j) = mean(deviatioVector);
        outSD(i,j) = sd(deviatioVector);
      }
    }
  }
  return List::create(Named("avg") = outAverage, Named("sd") = outSD);
}

double HiCSContrastC(NumericMatrix indexMap, NumericVector subspace, double alpha, int numRuns){
  double result = 0;
  for(int i=0; i<numRuns; i++){
    int referenceDim = subspace[std::rand() % (subspace.size())];
    result += KSTestC(randomSubspaceSliceC(indexMap, subspace, alpha, referenceDim), indexMap.column(referenceDim - 1));
  }
  return result/numRuns;
}

NumericMatrix deviationMatrixC(NumericMatrix indexMap, double alpha, int numRuns){
  NumericMatrix out(indexMap.ncol(), indexMap.ncol());
  colnames(out) = colnames(indexMap);
  rownames(out) = colnames(indexMap);

  for(int i=0; i< out.ncol(); i++){
    for(int j=0; j< out.ncol(); j++){
      if(i==j){
        out(i,j) = 0;
      }
      else{
        NumericVector subspace = NumericVector::create(i+1,j+1);
        out(i,j) = averageDeviationC(indexMap, subspace, alpha, i+1, numRuns);
      }
    }
  }
  return out;
}

NumericMatrix HiCSMatrixC(NumericMatrix indexMap, double alpha, int numRuns){
  NumericMatrix out(indexMap.ncol(), indexMap.ncol());
  colnames(out) = colnames(indexMap);
  rownames(out) = colnames(indexMap);

  for(int i=0; i< out.ncol(); i++){
    for(int j=i; j< out.ncol(); j++){
      if(i==j){
        out(i,j) = 0;
      }
      else{
        NumericVector subspace = NumericVector::create(i+1,j+1);
        out(i,j) = HiCSContrastC(indexMap, subspace, alpha, numRuns);
        out(j,i) = out(i,j);
      }
    }
  }
  return out;
}

File: R/deviationFunctions.R

deviation <- function(indexMap, subspace, referenceDim, alpha, numRuns){
  .Deprecated("deviationC")
  res <- 0
  for(i in 1:numRuns){
    res = res + KSTest(randomSubspaceSlice(indexMap, subspace, alpha, referenceDim), indexMap[,referenceDim, with=F][[1]])
  }
  res/numRuns
}

KSTest <- function(indexSelection, indexMapReferenceAttribute){
  .Deprecated("KSTestC")
  cumDistSubspace <- 0
  cumDistOriginal <- 0
  maxDistance <- 0
  numRemaining <- sum(indexSelection,na.rm = TRUE)
  numTotal <- length(indexSelection)

  for(i in 1:length(indexMapReferenceAttribute)){
    cumDistOriginal = (i+1)/numTotal
    if(indexSelection[indexMapReferenceAttribute[i]]){
      cumDistSubspace = cumDistSubspace + 1/numRemaining
    }
    maxDistance <- max(maxDistance, abs(cumDistSubspace - cumDistOriginal))
  }
  maxDistance
}

#' Index map for data set
#'
#' The index map allows to create subspace slices without working on the underlying data set but rather on the ordered index.
#'
#' @param dt data.table
#'
#' @return A data.table is returned that consists of indexes for the order of each column. E.g. a data table with the first column (0.3,0.5, 0.1) would get an index of (2,3,1), i.e. the elements sorted by (2,3,1) are in increasing order. The index starts with 1.
#' @export
sortedIndex <- function(dt){
  res <- copy(dt)
  res[,idx:=.I]
  l <- list()
  for (j in names(dt)){
    setorderv(res,c(j))
    l[[j]] <- res[,.(idx)][[1]]
  }
  data.table(do.call(cbind, l))
}

#' Index matrix for data set
#'
#' This is a wrapper for \code{\link{sortedIndex}}.
#'
#' @param dt data.table
#'
#' @return \code{\link{sortedIndex}} converted to a matrix
#' @export
sortedIndexMatrix <- function(dt){
  res <- as.matrix(sortedIndex(dt))
  colnames(res) <- colnames(dt)
  res
}

randomSubspaceSlice <- function(indexMap, subspace, alpha, referenceDim){
  .Deprecated("randomSubspaceSliceC")
  sliceSize <- ceiling(nrow(indexMap) * (alpha ^ (1/(length(subspace)-1))))
  selectionAll <- rep(T,nrow(indexMap))

  for(dim in subspace){
    if(dim != referenceDim){
      l <- ceiling(runif(1,1,nrow(indexMap)-sliceSize))
      r <- l + sliceSize
      selectionAll[indexMap[-(l:r), dim, with=F][[1]]] <- F
    }
  }
  selectionAll
}
